util/src/main/java/com/google/cloud/hadoop/util/CredentialFactory.java

```diff
@@ -254,131 +250,92 @@ public static Credential getCredentialFromMetadataServiceAccount()
-        tokenServerUrl);
+        options.getTokenServerUrl());
   }
 
   /**
    * Get credentials listed in a JSON file.
    *
-   * @param serviceAccountJsonKeyFile A file path pointing to a JSON file containing credentials.
    * @param scopes The OAuth scopes that the credential should be valid for.
    * @param transport The HttpTransport used for authorization
    */
-  private static Credential getCredentialFromJsonKeyFile(
-      String serviceAccountJsonKeyFile,
-      List<String> scopes,
-      HttpTransport transport,
-      String tokenServerUrl)
-      throws IOException, GeneralSecurityException {
-    logger.atFine().log("getCredentialFromJsonKeyFile(%s, %s)", serviceAccountJsonKeyFile, scopes);
+  private Credential getCredentialFromJsonKeyFile(List<String> scopes, HttpTransport transport)
+      throws IOException {
+    logger.atFine().log(
+        "getCredentialFromJsonKeyFile(%s, %s) from '%s'",
+        scopes, transport, options.getServiceAccountJsonKeyFile());
 
-    try (FileInputStream fis = new FileInputStream(serviceAccountJsonKeyFile)) {
+    try (FileInputStream fis = new FileInputStream(options.getServiceAccountJsonKeyFile())) {
       return GoogleCredentialWithRetry.fromGoogleCredential(
           GoogleCredential.fromStream(fis, transport, JSON_FACTORY).createScoped(scopes),
-          tokenServerUrl);
+          options.getTokenServerUrl());
     }
   }
 
-  private static Credential getCredentialsFromSAParameters(
-      String privateKeyId,
-      String privateKeyPem,
-      String serviceAccountEmail,
-      List<String> scopes,
-      HttpTransport transport,
-      String tokenServerUrl)
+  private Credential getCredentialsFromSAParameters(List<String> scopes, HttpTransport transport)
       throws IOException {
-    logger.atFine().log(
-        "getServiceAccountCredentialFromHadoopConfiguration(%s)", serviceAccountEmail);
-    if (serviceAccountEmail == null || privateKeyPem == null || privateKeyId == null) {
-      throw new IOException(
-          "Error reading service account credential from stream, "
-              + "expecting, 'client_email', 'private_key' and 'private_key_id'.");
-    }
-    PrivateKey privateKey = privateKeyFromPkcs8(privateKeyPem);
+    logger.atFine().log("getCredentialsFromSAParameters(%s, %s)", scopes, transport);
     GoogleCredential.Builder builder =
         new GoogleCredential.Builder()
             .setTransport(transport)
             .setJsonFactory(JSON_FACTORY)
-            .setServiceAccountId(serviceAccountEmail)
+            .setServiceAccountId(options.getServiceAccountEmail())
             .setServiceAccountScopes(scopes)
-            .setServiceAccountPrivateKey(privateKey)
-            .setServiceAccountPrivateKeyId(privateKeyId);
-    return new GoogleCredentialWithRetry(builder, tokenServerUrl);
+            .setServiceAccountPrivateKey(privateKeyFromPkcs8(options.getServiceAccountPrivateKey()))
```

Comment:
why not move privateKeyFromPkcs8 into getServiceAccountPrivateKey?

Reply:
Options class responsible only for holding and validating user configuration - transformation is better done in the place where it needed, because there could be other use cases that don't need this transformation.